### PR TITLE
gh-117613: Argument Clinic: ensure that `defining_class` params are positional-only

### PR DESCRIPTION
**Pull request by neonene**
This ensures that the AC regards `defining_class` as positional only, so that `_PyArg_UnpackKeywords()` can avoid a crash when a keyword mismatches.  The crash is not reproducible in `test_clinic.c`, as it defines `Py_BUILD_CORE_MODULE`.

https://github.com/python/cpython/blob/25f6ff5d3e92305659db62e7f7545f823f0dbd05/Python/getargs.c#L2515-L2526

* Issue: gh-117613

### CODE REVIEW

**Comment by erlend-aasland** [comment]
> The crash is not reproducible in test_clinic.c, as it defines Py_BUILD_CORE_MODULE.

We also have `Modules/_testclinic_limited.c` which explicitly does not define `Py_BUILD_CORE_MODULE`.

**Comment by neonene** [comment]
I failed to make a test.  I should have said that `Py_BUILD_CORE` needs to be defined. Also, it seems to me that the current AC does not support yet `METH_FASTCALL`, `NUM_KEYWORDS`, `KWTUPLE`, if `limited_capi` is true in `clanguage.py`.

**Comment by erlend-aasland** [comment]
> I should have said that Py_BUILD_CORE needs to be defined.

In that case, you can use `Modules/_testclinic.c` which explicitly defines `Py_BUILD_CORE_MODULE`

**Comment by neonene** [comment]
As long as the AC generates the check as `#if defined(Py_BUILD_CORE) && !defined(Py_BUILD_CORE_MODULE)`, both `_testclinic` and `_testclinic_limited` would not reproduce a crash.  And the check should not be changed in any test cases, right? Then, is there a good built-in module to confirm this issue, or `_testclinic_builtin` is needed?  No reproducer is fine with me, though.

cc @encukou @vstinner:  The issue, I think, is that the AC can make a keyword tuple whose size has an extra count of a `defining_class` parameter, in the `clinic/*.c.h` file of a builtin module such as `_datetime`.

**Comment by vstinner** [comment]
> Tests / Check if generated files are up to date (pull_request) Failing after 2m 

You can run `make clinic && make clinic-tests` and commit changed files.

**Comment by vstinner** [comment]
I propose https://github.com/python/cpython/pull/117896 to enhance `@defining_class` tests.

**Review by vstinner — APPROVED** [review]
LGTM.

@erlend-aasland: Do you want to double check?

File: Lib/test/test_clinic.py [hunk 1]
```diff
@@ -2508,6 +2508,15 @@ def test_state_func_docstring_only_one_param_template(self):
         """
         self.expect_failure(block, err, lineno=7)
 
+    def test_kind_defining_class(self):
+        function = self.parse_function("""
+            module m
+            m.func
+                cls: defining_class
```
Comment (erlend-aasland) [inline Lib/test/test_clinic.py]: This test seems strange to me. IMO, it should not be allowed to use the `defining_class` converter if it is a module level function; a class method should be required.

Comment (vstinner) [reply Lib/test/test_clinic.py]: I merged my PR https://github.com/python/cpython/pull/117896 which enhance `@defining_class` tests. There are now 3 functional tests, it's maybe enough?

Comment (erlend-aasland) [reply Lib/test/test_clinic.py]: Yes, I think that is enough.

Comment (neonene) [reply Lib/test/test_clinic.py]: I think I fixed my code.  I'll post a PR after checking how to disallow `defining_class` placed on a module input.

**Comment by vstinner** [comment]
You should rebase your PR on main and re-run `make clinic && make clinic-tests`.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
@vstinner, I think we should backport this to 3.12.

**Comment by neonene** [comment]
I think I'm not finished rerunning `clinic.py`.

**Comment by erlend-aasland** [comment]
> I think I'm not finished rerunning `clinic.py`.

No problem; auto-merge will only work if the CI is green :)

**Comment by erlend-aasland** [comment]
Thanks, @neonene! I synced with `main`, just to be on the safe side; when dealing with checked-in generated code it's best to be completely up to date when merging.

**Comment by neonene** [comment]
Thanks for reviewing and pointing out my mistakes, @vstinner @erlend-aasland.

**Comment by erlend-aasland** [comment]
@neonene, do you want to do the backport?

**Comment by vstinner** [comment]
Thanks @neonene for this nice fix. I never noticed that generated code was wrong.

**Pull request merged**